baseservice/iodevices/input_devices.py

```diff
@@ -0,0 +1,399 @@
+from abc import ABCMeta, abstractmethod
+from threading import Event
+from time import time, sleep
+from typing import Optional, Tuple, List, Dict, Union
+
+from baseservice.iodevices.common import Message, DeviceHeaders
+from baseservice.utils import KwargsException, StatefulListIterator
+
+
+class InputDeviceException(KwargsException):
+    """
+    a base exception class for all input device related exceptions
+    """
+    pass
+
+
+class InputTransaction(metaclass=ABCMeta):
+    """
+    this is the base class for input transaction objects.
+
+    after reading a message from the device (using 'with_transaction=True') the reader should commit/rollback
+    the transaction, to signal the device if the message is done processing or not
+    """
+
+    def __init__(self, device: 'InputDevice'):
+        """
+
+        :param device: the input device that returned that transaction
+        """
+        self._device: 'InputDevice' = device
+        self._finished: Event = Event()
+
+    @property
+    def device(self) -> 'InputDevice':
+        """
+        :return: the input device that returned that transaction
+        """
+        return self._device
+
+    @property
+    def finished(self) -> bool:
+        """
+        :return: 'True' if the transaction committed/rolledback, 'False' otherwise
+        """
+        return self._finished.is_set()
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.rollback()  # TODO: should a transaction be auto-committed if no exception was raised?
+
+    def wait_for_finish(self, timeout: Optional[float] = None) -> bool:
+        """
+        waits for the transaction to finish (commit/rollback).
+        it timeout is not None, after timeout seconds, it will return self.finished
+
+        :param timeout: the timeout (in seconds) to wait for the transaction to finish
+
+        :return: the value of self.finished
+        """
+        return self._finished.wait(timeout=timeout)
+
+    def commit(self):
+        """
+        commits the transaction
+        """
+        if self.finished:
+            return
+        self._commit()
+        self._finished.set()
+
+    def rollback(self):
+        """
+        rolls back the transaction
+        """
+        if self.finished:
+            return
+        self._rollback()
+        self._finished.set()
+
+    @abstractmethod
+    def _commit(self):
+        """
+        this method should be implemented by child classes to actually perform the commit
+        """
+        pass
+
+    @abstractmethod
+    def _rollback(self):
+        """
+        this method should be implemented by child classes to actually perform the rollback
+        """
+        pass
+
+
+class WrapperTransaction(InputTransaction):
+    """
+    wraps a transaction for an underlying device. used for wrapper devices.
+    """
```

Comment:
Hard to understand the documentation. Consider rephrasing it

Reply:
care to make a suggestion?
maybe it's hard to understand, because there's only one wrapper device so far?